spec/abstract/AARCH64/ArchDecode_A.thy

```diff
@@ -0,0 +1,334 @@
+      then if length args \<ge> 4 \<and> length cps \<ge> 1
+        then let irq_word = args ! 0;
+                 trigger = args ! 1;
+                 index = args ! 2;
+                 depth = args ! 3;
+                 cnode = cps ! 0;
+                 irq = ucast irq_word
+        in doE
+          arch_check_irq irq_word;
+          irq_active \<leftarrow> liftE $ is_irq_active irq;
+          whenE irq_active $ throwError RevokeFirst;
+          dest_slot \<leftarrow> lookup_target_slot cnode (data_to_cptr index) (unat depth);
+          ensure_empty dest_slot;
+          returnOk $ ARMIRQControlInvocation irq dest_slot src_slot (trigger \<noteq> 0)
+        odE
+      else throwError TruncatedMessage
+    else throwError IllegalOperation)"
+
+definition attribs_from_word :: "machine_word \<Rightarrow> vm_attributes" where
+  "attribs_from_word w \<equiv> {attr.  \<not>w!!0 \<and> attr = Execute \<or> \<not>w !! 2 \<and> attr = Device}"
+
+definition make_user_pte :: "paddr \<Rightarrow> vm_attributes \<Rightarrow> vm_rights \<Rightarrow> vmpage_size \<Rightarrow> pte" where
+  "make_user_pte addr attr rights vm_size \<equiv>
+     PagePTE addr (vm_size = ARMSmallPage) (attr - {Global}) rights"
+
+definition check_vspace_root :: "arch_cap \<Rightarrow> nat \<Rightarrow> (obj_ref \<times> asid, 'z) se_monad" where
+  "check_vspace_root cap arg_no \<equiv>
+     case cap of
+       PageTableCap pt True (Some (asid, _)) \<Rightarrow> returnOk (pt, asid)
+     | _ \<Rightarrow> throwError $ InvalidCapability arg_no"
+
+type_synonym 'z arch_decoder =
+  "data \<Rightarrow> data list \<Rightarrow> cslot_ptr \<Rightarrow> arch_cap \<Rightarrow> (cap \<times> cslot_ptr) list \<Rightarrow>
+    (arch_invocation,'z) se_monad"
+
+definition decode_fr_inv_map :: "'z::state_ext arch_decoder" where
+  "decode_fr_inv_map label args cte cap extra_caps \<equiv> case cap of
+     FrameCap p R pgsz dev mapped_address \<Rightarrow>
+       if length args > 2 \<and> length extra_caps > 0
+       then let
+           vaddr = args ! 0;
+           rights_mask = args ! 1;
+           attr = args ! 2;
+           vspace_cap = fst (extra_caps ! 0)
+         in doE
+           (pt, asid) \<leftarrow> check_vspace_root cap 1;
+           pt' \<leftarrow> lookup_error_on_failure False $ find_vspace_for_asid asid;
+           whenE (pt' \<noteq> pt) $ throwError $ InvalidCapability 1;
+           check_vp_alignment pgsz vaddr;
+           pg_bits \<leftarrow> returnOk $ pageBitsForSize pgsz;
+           case mapped_address of
+             Some (asid', vaddr') \<Rightarrow> doE
+               whenE (asid' \<noteq> asid) $ throwError $ InvalidCapability 1;
+               whenE (vaddr' \<noteq> vaddr) $ throwError $ InvalidArgument 0
+             odE
+           | None \<Rightarrow> doE
+               vtop \<leftarrow> returnOk $ vaddr + mask (pageBitsForSize pgsz);
+               whenE (vtop \<ge> user_vtop) $ throwError $ InvalidArgument 0
```

Comment:
Nicely spotted, it looks like Haskell is off by one. Generally `<=` is nicer for overflow reasoning, so I think we should leave ASpec as is, but use the same formulation between C and Haskell, so that the conversion happens in Refine, not CRefine, and we have the nicer version in AInvs. Unless we want to adjust the C, which would make it nicer everywhere. I might have a look at that, actually.

Reply:
actually, I think it is fine as is. In Haskell, the `-1` is already part of the definition of `vtop`, and the parentheses are set to make the proof to the <= in ASpec more obvious. Will we find out for sure which way it is in the AInvs proofs..